### DISCUSSION TITLE
Unpremultiply Alpha

### DISCUSSION THREAD

**Discussion by Anti-Distinctlyminty** [ImageMagick/ImageMagick#5074] [Digital Image Processing]
I'm trying to do something like 
`magick montage -mode concatenate -tile 1.png 2.png 3.png -background "white" out.png`
Where `1.png` `2.png` `3.png` are have a premultiplied alpha.
However, the resulting out.png has dark fringing

![Resolve_VWz0OwRORE](https://user-images.githubusercontent.com/1378483/164965312-014b5e57-db8b-4b52-a397-6f1864be659e.png)

But as you can see in the attached images, if I unpremult the image in a compositing app then it looks fine

![Resolve_jhuy3ETpom](https://user-images.githubusercontent.com/1378483/164965310-11634c3f-2f93-4c68-85f2-6db3677d8ef8.png)

Can I unpremultiply the images when creating the montage?

**Comment by snibgo** [comment]
> Can I unpremultiply the images when creating the montage?

"Premultiply" means that before an image operation that mixes pixels from different images, the image pixels are multiplied by alpha. After the image operation, pixels are divided by alpha.

Montage doesn't mix pixels from multiple images, so I don't see how premultiply can make a difference.

But you can of course, multiply RGB values by alpha values, if you want:
```
magick in.png ( +clone -alpha extract ) -compose Multiply -composite out.png
```

**Reply by Anti-Distinctlyminty** [reply]
> "Premultiply" means that before an image operation that mixes pixels from different images, the image pixels are multiplied by alpha. After the image operation, pixels are divided by alpha.

I should have been more specific. The images come from a renderer which premultiplies the RGB channels of the output (with black I believe). So if I'm recompositing them on white (As I'm trying to do with the montage command above) I need the image divided by the alpha. I tried

```
magick in.png ( +clone -alpha extract ) -compose Divide -composite out.png
```
But I ended up with what looked like just the alpha. I'm assuming its the order of the images being fed into the composite, but I'm unsure how to switch them.

**Comment by snibgo** [comment]
I suggest you post your input images.

**Comment by Anti-Distinctlyminty** [comment]
Here's the relevant part of the image
![AlphaDivide](https://user-images.githubusercontent.com/1378483/164982402-2b1815bf-4702-41ab-9357-f4a2c0b2adbf.png)

**Comment by snibgo** [comment]
In your image, which I have called hair.png, the strands with transparency are darker than opaque hair. Perhaps the original photo (or 3D ray-tracer) had a dark background.

We can make pixels with transparency lighter by dividing all pixels by their alpha.

We can divide by the extracted alpha. This would remove the alpha, so we CopyOpacity to restore it. Windows BAT syntax.
```
magick ^
  hair.png ^
  ( +clone -alpha extract +write mpr:ALPH ) ^
  -compose DivideSrc -composite mpr:ALPH -compose CopyOpacity -composite ^
  x3.png
```
If it was made by a ray-tracer, it should be told to make a transparent background. If it makes an opaque background, and you know what that background is, the hair colours can be reconstructed accurately. See [Watermarks](http://im.snibgo.com/watermark.htm), which addresses exactly that problem,

**Comment by Anti-Distinctlyminty** [comment]
Thanks for the help @snibgo. Here's the original and the one processed as you suggested above:
![out](https://user-images.githubusercontent.com/1378483/164986723-33fe97a8-ae45-49c7-90d1-d141ea2eee5f.png)
It's definitely lighter but if we compare it to putting it though an alpha divide in Fusion:
![out_fusion](https://user-images.githubusercontent.com/1378483/164986779-957a84d5-de4c-42a4-a17a-bbb36ea97922.png)

It's getting there but there's still some darkening for some reason. I'm struggling to see what the difference is in the processing between the two.

**Comment by snibgo** [comment]
I don't know what Fusion does.

The method I showed makes the semi-transparent hair pixels roughly the same colour and lightness as nearby opaque hair colours. When flattened against white, they become even more lighter.

If you want them lighter still, doing the division in linear RGB space will do that:
```
magick ^
  hair.png ^
  ( +clone -alpha extract +write mpr:ALPH ) ^
  -colorspace RGB ^
  -compose DivideSrc -composite mpr:ALPH -compose CopyOpacity -composite ^
  -colorspace sRGB ^
  x5.png
```

**Comment by finscn** [comment]
with ImageMagick 7 : 
```
// premultiplied
convert  input.png -channel rgb -fx "u*u.a" output.png

// un-premultiplied
convert  input.png -channel rgb -fx "u/max(u.a, 1/1024)" output.png
```
